File: src/platform/lm4f/conf.py
# Configuration file for the LM4F microcontroller
import fnmatch
import glob
import os

comp.Append(CPPPATH = ['src/platform/%s/inc' % platform])
comp.Append(CPPPATH = ['src/platform/%s/driverlib' % platform])

# The microcontroller LM4F120H5QR supports USB CDC
comp.Append(CPPPATH = ['src/platform/%s/usblib' % platform])
comp.Append(CPPPATH = ['src/platform/%s/usblib/device' % platform])

fwlib_files = " ".join(glob.glob("src/platform/%s/driverlib/*.c" % platform))

specific_files = "startup_gcc.c platform.c platform_int.c lm4f_pio.c"

# The microcontroller LM4F120H5QR supports USB CDC
fwlib_files = fwlib_files + " " + " ".join(glob.glob("src/platform/%s/usblib/*.c" % platform))
fwlib_files = fwlib_files + " " + " ".join(glob.glob("src/platform/%s/usblib/device/*.c" % platform))
specific_files = specific_files + " usb_serial_structs.c"

ldscript = "lm4f.ld"

# Prepend with path
specific_files = fwlib_files + " " + " ".join( [ "src/platform/%s/%s" % ( platform, f ) for f in specific_files.split() ] )
specific_files += " src/platform/cortex_utils.s src/platform/arm_cortex_interrupts.c"
ldscript = "src/platform/%s/%s" % ( platform, ldscript )

comp.Append(CPPDEFINES = ["FOR" + comp[ 'cpu' ],'gcc'])
comp.Append(CPPDEFINES = ['CORTEX_M4'])

# TODO: check if we need to pass a flag to FP being active
# Standard GCC Flags
comp.Append(CCFLAGS = ['-ffunction-sections','-fdata-sections','-fno-strict-aliasing','-Wall'])
comp.Append(LINKFLAGS = ['-nostartfiles','-nostdlib','-T',ldscript,'-Wl,--gc-sections','-Wl,--allow-multiple-definition'])
comp.Append(ASFLAGS = ['-x','assembler-with-cpp','-c','-Wall','$_CPPDEFFLAGS'])
comp.Append(LIBS = ['c','gcc','m'])

# TODO: Consider using thumb2 instruction set. May need to deal with src/platform/ arm files.
TARGET_FLAGS = ['-mcpu=cortex-m4','-mthumb']

# Configure General Flags for Target
comp.Prepend(CCFLAGS = [TARGET_FLAGS,'-mlittle-endian'])
comp.Prepend(LINKFLAGS = [TARGET_FLAGS,'-Wl,-e,ResetISR','-Wl,-static'])
comp.Prepend(ASFLAGS = TARGET_FLAGS)

# Toolset data
tools[ 'lm4f' ] = {}

# Programming function
def progfunc_lm4f( target, source, env ):
  outname = output + ".elf"
  os.system( "%s %s" % ( toolset[ 'size' ], outname ) )
  print "Generating binary image..."
  os.system( "%s -O binary %s %s.bin" % ( toolset[ 'bin' ], outname, output ) )

tools[ 'lm4f' ][ 'progfunc' ] = progfunc_lm4f

File: src/platform/lm4f/conf.lua
-- Configuration file for the LM4F microcontroller

addi( sf( 'src/platform/%s/inc', platform ) )
addi( sf( 'src/platform/%s/driverlib', platform ) )

-- The microcontroller LM4F120H5QR supports USB CDC
addi( sf( 'src/platform/%s/usblib', platform ) )
addi( sf( 'src/platform/%s/usblib/device', platform ) )

specific_files = "startup_gcc.c platform.c platform_int.c lm4f_pio.c"
local fwlib_files = utils.get_files( "src/platform/" .. platform .. "/driverlib", ".*%.c$" )

-- The microcontroller LM4F120H5QR supports USB CDC
fwlib_files = fwlib_files .. " " .. utils.get_files( "src/platform/" .. platform .. "/usblib", ".*%.c$" ) 
fwlib_files = fwlib_files .. " " .. utils.get_files( "src/platform/" .. platform .. "/usblib/device", ".*%.c$" )
specific_files = specific_files .. "  usb_serial_structs.c"

ldscript = "lm4f.ld"

-- Prepend with path
specific_files = fwlib_files .. " " .. utils.prepend_path( specific_files, "src/platform/" .. platform )
specific_files = specific_files .. " src/platform/cortex_utils.s src/platform/arm_cortex_interrupts.c"
ldscript = sf( "src/platform/%s/%s", platform, ldscript )

addm{ "FOR" .. comp.cpu:upper(), 'gcc', 'CORTEX_M4' }

-- Standard GCC flags
addcf{ '-ffunction-sections', '-fdata-sections', '-fno-strict-aliasing', '-Wall' }
addlf{ '-nostartfiles', '-nostdlib', '-T', ldscript, '-Wl,--gc-sections', '-Wl,--allow-multiple-definition' }
addaf{ '-x', 'assembler-with-cpp', '-Wall' }
addlib{ 'c','gcc','m' }

-- TODO: Consider using thumb2 instruction set. May need to deal with src/platform/ arm files.
local target_flags =  {'-mcpu=cortex-m4','-mthumb' }

-- Configure general flags for target
addcf{ target_flags, '-mlittle-endian' }
addlf{ target_flags, '-Wl,-e,ResetISR', '-Wl,-static' }
addaf( target_flags )

-- Toolset data
tools.lm4f = {}

-- Array of file names that will be checked against the 'prog' target; their absence will force a rebuild
tools.lm4f.prog_flist = { output .. ".bin" }

-- We use 'gcc' as the assembler
toolset.asm = toolset.compile

